rust/noosphere-storage/src/partitioned.rs

```diff
@@ -0,0 +1,106 @@
+use crate::{Disposable, Store};
+use anyhow::Result;
+use async_stream::try_stream;
+use async_trait::async_trait;
+
+/// Maps all [Store] method calls with a key to a prefixed form.
+#[derive(Clone)]
+pub struct PartitionedStore<S>
+where
+    S: Store,
+{
+    store: S,
+    partition_key: Vec<u8>,
+    end_partition_key: Vec<u8>,
+}
+
+impl<S> PartitionedStore<S>
+where
+    S: Store,
+{
+    pub fn new(store: S) -> Self {
+        let prefix: Vec<u8> = format!("{:0<10}-", rand::random::<u32>()).into();
+        Self::with_partition_key(store, prefix)
+    }
+
+    pub fn with_partition_key(store: S, partition_key: Vec<u8>) -> Self {
+        let mut end_partition_key = partition_key.clone();
+        end_partition_key.push(u8::MAX);
+        Self {
+            store,
+            partition_key,
+            end_partition_key,
+        }
+    }
+
+    /// Returns `bool` indicating whether `key` is within
+    /// the partition key space.
+    fn partition_key_contains(&self, key: &[u8]) -> bool {
+        key.starts_with(self.partition_key.as_slice())
+    }
+
+    pub fn get_key_range(&self) -> (&Vec<u8>, &Vec<u8>) {
```

Comment:
The symbol for an "end" of the prefix range could be improved possibly with the `partition_key` with the last character incremented. Both Rocks and IDB use a partitioned implementation as well as have the same interface for key ranges. Rocks has a more robust solution here which we could adopt: https://docs.rs/rocksdb/0.21.0/rocksdb/struct.PrefixRange.html